cli/src/diff_util.rs

```diff
@@ -648,6 +667,67 @@ pub fn show_color_words_diff(
     Ok(())
 }
 
+pub fn show_file_by_file_diff(
+    ui: &Ui,
+    repo: &dyn Repo,
+    formatter: &mut dyn Formatter,
+    tool: &ExternalMergeTool,
+    tree_diff: TreeDiffStream,
+    path_converter: &RepoPathUiConverter,
+) -> Result<(), DiffRenderError> {
+    fn create_file(
+        path: &RepoPath,
+        wc_dir: &Path,
+        value: MaterializedTreeValue,
+    ) -> Result<PathBuf, DiffRenderError> {
+        let fs_path = path.to_fs_path(wc_dir);
+        std::fs::create_dir_all(fs_path.parent().unwrap())?;
+        let content = diff_content(path, value)?;
+        let mut buffer = File::create(&fs_path)?;
+        buffer.write_all(&content.contents)?;
+        Ok(fs_path)
+    }
+
+    let temp_dir = new_utf8_temp_dir("jj-diff-")?;
+    let left_wc_dir = temp_dir.path().join("left");
+    let right_wc_dir = temp_dir.path().join("right");
+    let mut diff_stream = materialized_diff_stream(repo.store(), tree_diff);
+    async {
+        while let Some((path, diff)) = diff_stream.next().await {
+            let ui_path = path_converter.format_file_path(&path);
+            let (left_value, right_value) = diff?;
+
+            match (&left_value, &right_value) {
+                (_, MaterializedTreeValue::AccessDenied(source))
+                | (MaterializedTreeValue::AccessDenied(source), _) => {
+                    write!(
+                        formatter.labeled("access-denied"),
+                        "Access denied to {ui_path}:"
+                    )?;
+                    writeln!(formatter, " {source}")?;
+                    continue;
+                }
+                _ => {}
+            }
+            let left_path = create_file(&path, &left_wc_dir, left_value)?;
+            let right_path = create_file(&path, &right_wc_dir, right_value)?;
+
```

Comment:
I would print some sort of heading here. I think that *at least* a newline is called for. But perhaps we should do something nicer:

> (a newline)
> JJ: Comparing XXX and YYY

Either JJ:, or the whole line, or the whole line except for the paths could be in yellow.

More questionably, in the common case you could print:

> Comparing {left,right}/PATH

Ideally, the heading would become templated at some point.

This is also important if the diff is opened in a separate window and `jj` waits for the program to finish. (Guess on Windows, people will just get 15 windows at once unless they take special care).

Reply:
Also, adding `#5 of 20` would be helpful.

Reply:
I added this, but I am not sure that I actually like it.  I will go with whatever the crowd decides here though

Reply:
If we print header by ourselves, I would copy the one from color-words/git diffs. I'm not sure if the header should be emitted by default. The diff command usually prints some header, so it's redundant if the tool output is streamed.

Reply:
I elaborated a bit in https://github.com/martinvonz/jj/pull/3982#discussion_r1660300300. If the program just outputs some text and prints some header, I'd agree. If it's more like `code --wait`, I'm not so sure.

Reply:
Maybe that's another reason we'll need `type = stream|gui|tui`.

Reply:
stripped out headers for now and will offer a separate PR just for headers